pkg/controllers/provisioning/v1alpha1/reallocation/terminate.go

```diff
@@ -30,39 +30,46 @@ type Terminator struct {
 	cloudprovider cloudprovider.Factory
 }
 
-func (t *Terminator) CordonNodes(ctx context.Context, nodes []*v1.Node) error {
+// AddTTLs takes in a list of underutilized nodes and adds TTL to them
+func (t *Terminator) AddTTLs(ctx context.Context, nodes []*v1.Node) error {
 	for _, node := range nodes {
-		if node.Spec.Unschedulable {
-			continue
-		}
-		persisted := node.DeepCopyObject()
-		node.Spec.Unschedulable = true
+		persisted := node.DeepCopy()
+		node.Annotations[v1alpha1.ProvisionerTTLKey] = time.Now().Add(60 * time.Second).Format(time.RFC3339)
 		if err := t.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
-			return fmt.Errorf("patching node %s, %w", node.Name, err)
+			return fmt.Errorf("patching node, %w", err)
 		}
+		zap.S().Debugf("Added TTL to nodes %s", node.Name)
 	}
 	return nil
 }
 
-func (t *Terminator) AddTTLs(ctx context.Context, nodes []*v1.Node) error {
+// CordonNodes takes in a list of expired nodes as input and cordons them
+func (t *Terminator) CordonNodes(ctx context.Context, nodes []*v1.Node) error {
 	for _, node := range nodes {
-		if _, ok := node.Annotations[v1alpha1.ProvisionerTTLKey]; !ok {
-			persisted := node.DeepCopy()
-			node.Annotations[v1alpha1.ProvisionerTTLKey] = time.Now().Add(300 * time.Second).Format(time.RFC3339)
-			if err := t.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
-				return fmt.Errorf("patching node, %w", err)
-			}
+		persisted := node.DeepCopy()
+		node.Spec.Unschedulable = true
+		if err := t.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			return fmt.Errorf("patching node %s, %w", node.Name, err)
 		}
+		zap.S().Debugf("Cordoned node %s", node.Name)
 	}
 	return nil
 }
 
+// DeleteNodes will use a cloudprovider implementation to delete a set of nodes
 func (t *Terminator) DeleteNodes(ctx context.Context, nodes []*v1.Node, spec *v1alpha1.ProvisionerSpec) error {
-	err := t.cloudprovider.CapacityFor(spec).Delete(ctx, nodes)
-	if err != nil {
+	// 1. Delete node in cloudprovider's instanceprovider
+	if err := t.cloudprovider.CapacityFor(spec).Delete(ctx, nodes); err != nil {
 		return fmt.Errorf("terminating %d cloudprovider instances, %w", len(nodes), err)
 	}
-	zap.S().Infof("Terminating %d instances", len(nodes))
 
+	// 2. Delete node in APIServer to ensure
+	// TODO: Prevent leaked nodes: ensure a node is not deleted in apiserver if not deleted in cloudprovider
```

Comment:
Just curious - what more work are you envisioning here? Right now it looks like the code is pretty conservative in that it only calls `kubeClient.Delete` if the cloudprovider didn't return an error?

Reply:
The work would be to use the returned ids from the cloudprovider's specific deleteNode function, and then only delete those ids in the apiserver.

Reply:
Makes sense